Adding media query to support responsive design

0916dhkim opened the pull request:
## Specify project
Client

## Description

Sign-in screen is not responsive, and too wide for high-resolution displays.
Set maximum width of the scroll view to be 800.

### original
![original](https://user-images.githubusercontent.com/8978815/90477585-9de27800-e166-11ea-82e1-ceb9ce269bf7.png)

### fixed
![fixed](https://user-images.githubusercontent.com/8978815/90477607-a8047680-e166-11ea-9b50-6877d1128b50.png)

## Checklist

Before you create this PR confirms that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide](https://github.com/dooboolab/hackatalk/blob/master/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] Run `yarn lint && yarn tsc`
- [x] Run `yarn test` or `yarn test -u` if you need to update snapshot.
- [x] I am willing to follow-up on review comments in a timely manner.

0916dhkim commented:
> Well I think this should be handled by the media query in the future.

Yes, especially because we need responsiveness for other parts of Hackatalk, too.

0916dhkim commented:
I will work on adding media queries to styles.

0916dhkim commented:
@hyochan It seems like React-Native does not support media queries out of the box. What do you think about using this library?
https://github.com/wcandillon/react-native-responsive-ui

0916dhkim commented:
@hyochan I made a hook named `useResponsiveDesign` which returns `'small'` `'medium'` or `'large'` depending on screen width. I did not add any styled component in `SignIn` component because we're not re-using the component, but if we do need to implement an re-usable responsive container component, using `useResponsiveDesign` hook with styled component will be reasonably straight-forward.

hyochan commented:
@0916dhkim What about `mobile`, `tablet`, `desktop` as a naming?
I'd like to figure out how this can nicely join with `stlyed-component` so we can use it not exactly but similar to https://github.com/dooboolab/dooboo-frontend-ts/blob/8921ee4c8dfbf3814fa342c0c3512897d250e657/src/components/screen/Intro.tsx#L27-L43.

0916dhkim commented:
@hyochan I like your naming better. I will see if I can come up with an implementation similar to your example.

0916dhkim commented:
@hyochan What do you think about writing media queries this way?
```ts
const StyledScrollView = styled.ScrollView<{ media: Media }>`
  alignSelf: center;
  width: 100%;
  ${({ media }) => media === 'tablet' && css`
    width: 75%;
  `}
  ${({ media }) => media === 'desktop' && css`
    width: 50%;
    max-width: 800;
  `}
`;
```

hyochan commented on client/src/components/screen/SignIn.tsx:
```diff
@@ -330,7 +344,7 @@ function SignIn(props: Props): ReactElement {
       <StatusBar />
 
-      <ScrollView style={{ alignSelf: 'stretch' }}>
+      <StyledScrollView media={media}>
```
Everything looks fabulous except that we should put this `media` props here every time!

hyochan commented:
@0916dhkim How do you like [this update](https://github.com/dooboolab/hackatalk/pull/209/commits/6870205d9f14e2ac48e4510d9943d3b517ea400c)?

0916dhkim commented on client/src/utils/media.ts:
```diff
@@ -0,0 +1,26 @@
+  mobile: boolean,
+}
+
+const isDesktop = (): boolean => useMediaQuery({ minWidth: 1224 });
+const isTablet = (): boolean => useMediaQuery({ minWidth: 768 });
+const isMobile = (): boolean => useMediaQuery({ minWidth: 0 });
+
+export const getMediaQuery = (): MediaQueryType => {
+  return {
+    desktop: isDesktop(),
+    tablet: isTablet(),
+    mobile: isMobile(),
+  };
+};
```
```suggestion
export const useMedia = (): MediaQueryType => {
  const isDesktop = useMediaQuery({
    minWidth: 1224
  });
  const isTablet = useMediaQuery({
    minWidth: 768,
    maxWidth: 1223,
  });
  const isMobile = useMediaQuery({
    minWidth: 0,
    maxWidth: 767,
  });
  return {
    desktop: isDesktop,
    tablet: isTablet,
    mobile: isMobile,
  };
};
```

hyochan replied:
@0916dhkim Nice updates! I just want to point out here that if we only want to consider `mobile` and the `tablet` or the bigger screens, could we avoid using `desktop` in our implementation?

```
const StyledScrollView = styled.ScrollView`
  align-self: center;

  ${({ theme: { mobile } }) => mobile && css`
    width: 100%;
  `}

  ${({ theme: { tablet } }) => tablet && css`
    width: 50%;
    max-width: 800;
  `}
`;
```

hyochan replied:
I am afraid of using `minWidth` and `maxWidth` strictly, since the above isn't possible currently.

0916dhkim commented on client/src/App.tsx:
```diff
@@ -28,6 +28,7 @@ import { Image } from 'react-native';
 import { LoadingIndicator } from 'dooboo-ui';
 import RootNavigator from './components/navigation/RootStackNavigator';
 import { User } from 'types/graphql';
+import { getMediaQuery } from './utils/media';
```
```suggestion
import { useMedia } from './utils/media';
```

0916dhkim commented on client/src/App.tsx:
```diff
@@ -125,9 +126,20 @@ function App(): ReactElement {
 
 const HackatalkThemeProvider: FC<{ children: ReactElement }> = ({ children }) => {
   const colorScheme = useColorScheme();
+  const mediaQuery = getMediaQuery();
```
```suggestion
  const mediaQuery = useMedia();
```

0916dhkim reviewed (commented):
I think this update is more flexible way of doing media queries.

I have modified the `getMediaQuery` because it violates the "Rules of Hooks"
(A hook can only be used inside another hook or function component)
https://reactjs.org/docs/hooks-rules.html

0916dhkim commented:
@hyochan I have simplified the code.
> could we avoid using desktop in our implementation?

I removed `mobile` instead of `desktop` because I think we are taking mobile-first approach here and developers can provide overrides for `desktop`.

> I am afraid of using minWidth and maxWidth strictly, since the above isn't possible currently.

I changed `useMediaQuery` to only use `maxWidth`, however, I tried to make `tablet` and `desktop` mutually exclusive (as it would be weird to have both styles applied).
```ts
  return {
    tablet: isTablet && !isDesktop,
    desktop: isDesktop,
  };
```

hyochan commented on client/src/utils/media.ts:
```diff
@@ -0,0 +1,17 @@
+  const isTablet = useMediaQuery({ minWidth: 768 });
+  const isDesktop = useMediaQuery({ minWidth: 1224 });
+  return {
+    tablet: isTablet && !isDesktop,
+    desktop: isDesktop,
```
Can we have `mobile` and `tablet` here? I'd like to consider `mobile` as our first citizen and rests as `> tablet` in first hand!

0916dhkim replied:
> I'd like to consider mobile as our first citizen

@hyochan That's exactly why I chose `tablet` and `desktop` and left out `mobile` here.
```ts
const StyledScrollView = styled.ScrollView`

  width: 100%;  // Styles are for mobile by default (first-citizen)

  // Override for tablet
  ${({ theme: { tablet } }) => tablet && css`
    width: 75%;
  `}

  // Override for desktop
  ${({ theme: { desktop } }) => desktop && css`
    width: 50%;
    max-width: 800;
  `}
`;
```

hyochan replied:
What if we have only `tablet` here? This isn't `> tablet` since I see `isTablet && !isDesktop`.
Originally, media query overrides other bigger size screens if it reaches out last big screen which here would be the `tablet`.

hyochan replied:
Also, I want to order `rebasing` since there is a conflict :)

hyochan reviewed (approved):
Great work on this!!

Additionally, it looks like we have similar approaches as in [media query in material-ui](https://material-ui.com/components/use-media-query).
